### PR TITLE
Delete with devfile.yaml

### LINKED ISSUE #5478: `odo delete component` -  in the directory with `devfile.yaml`
/kind task

## Acceptance Criteria 
- [ ] If `odo delete component` is executed in a directory with `devile.yaml` it should delete all cluster resources related to that devfile (both inner loop and outer loop)
- [ ] Command to delete component should print what will be deleted first and ask the user to confirm.
- [ ] Command should have a flag to "force" deletion without confirmation
- [ ] No local files are deleted or modified.

### PR DESCRIPTION
**Pull request by valaparthvi**
Signed-off-by: Parthvi Vala <pvala@redhat.com>

**What type of PR is this:**
/kind feature

**What does this PR do / why we need it:**
This PR adds feature to delete component when devfile.yaml is accessible.

**Which issue(s) this PR fixes:**

Fixes #5478 

**PR acceptance criteria:**

- [x] Unit test 

- [x] Integration test 

- [ ] Documentation 

- [ ] Demo

**How to test changes / Special notes to the reviewer:**
1. Have a directory with devfile and source files
2. `odo push` or `odo dev`
3. `export PODMAN_CMD=echo` (optional)
4. `odo deploy`
5. `odo delete component`
Sample Output - 
```sh
➜  odo delete component 
The following components will be deleted: 
 •  Devfile Component: nodejs-prj1-api-abhz
 •  Kubernetes Resources:
         Deployment : my-component
         Deployment : my-component-2
? Are you sure you want to delete these resources? Yes
 ✓  Deleting Devfile component [22ms]
 ✓  Deleting Kubernetes resources [49ms]
 ✓  Successfully deleted component
```
TODO:
- [x] Currently no error is printed if the resource is not found on the cluster, it can be seen when run with debug level set between 1-4.
- [ ] Cleanup v2delete
- [x] Unit tests
- [ ] Documentation

### CODE REVIEW

File: pkg/component/delete/delete.go
Comment (feloy) [inline pkg/component/delete/delete.go:29]: Do you want to move this package-level function `component.Delete` to this `delete` package? It would be the occasion to cleanup the `component` package, and to see what to do with the `spinner`: do we want to rely on spinner into this layer, or only on the CLI one?

**Review by feloy — CHANGES_REQUESTED** [review]

File: pkg/component/delete/delete.go
Comment (valaparthvi) [reply pkg/component/delete/delete.go:29]: CLI does seem to be the ideal place to put spinner logic because it prints things to the terminal and is user facing, but then it is much more easier to place the spinner at the right place in the business logic.
But then we also discussed that business layer should just return an object and the CLI layer will print that object in JSON or human readable format.
I think we should stick to spinners in the CLI layer (write the methods in business logic in small pieces such that spinners at the CLI layer make sense), and use klog in the business layer for better understandably.

What do you think?

File: pkg/component/delete/delete.go
Comment (feloy) [inline pkg/component/delete/delete.go:31]: You can restrict the result with some devfile option: 
```
common.DevfileOptions{
		ComponentOptions: parsercommon.ComponentOptions{ComponentType: devfile.KubernetesComponentType},
}
```

File: pkg/component/delete/delete.go
Comment (feloy) [inline pkg/component/delete/delete.go:38]: You could move the `GetK8sComponentAsUnstructured` function to some devfile-related package (libdevfile?) as it is 100% related to devfile

File: pkg/odo/cli/delete/component.go
Comment (feloy) [inline pkg/odo/cli/delete/component.go:88]: As there is no return statement before the `spinner.End(true)`, you don't need this `defer`

File: pkg/odo/cli/delete/component.go
Comment (feloy) [inline pkg/odo/cli/delete/component.go:97]: Same here

**Review by feloy — CHANGES_REQUESTED** [review]

File: pkg/odo/cli/delete/component.go
Comment (feloy) [reply pkg/odo/cli/delete/component.go]: sorry, name == ""

File: pkg/component/delete/delete.go
Comment (feloy) [inline pkg/component/delete/delete.go:76]: Thsi changes the behaviour described in the comment line 91/67. Are you sure you want to change this behaviour?

File: pkg/component/delete/undeploy_handler.go
Comment (feloy) [inline pkg/component/delete/undeploy_handler.go:42]: This is dangerous to change this behaviour. Because of `ownerReferences` and the garbage collection triggered by the deletion of a previous resource, this resource could have been deleted in the meantime, and we don't want to fail the command at this point in this case. Or the user could have deleted this resource by hand, and wants to delete the other ones with `odo delete`, etc

Comment (valaparthvi) [reply pkg/component/delete/undeploy_handler.go:42]: Are you suggesting that in the case that the resource is not found for whatsoever reason, and this method returns a *NotFoundError, we return a `nil`?

Comment (feloy) [reply pkg/component/delete/undeploy_handler.go:42]: Yes, because we don't want to fail the complete command if the deletion of a single resource fails

File: pkg/component/delete/delete.go
Comment (valaparthvi) [reply pkg/component/delete/delete.go:76]: Can't believe I missed that. I was focused on removing log.Status out of this pkg, but also wanted to ensure the correct error messages are printed, I forgot about the effect of this one. Thanks for catching this. :100:

**Comment by valaparthvi** [comment]
/retest
Failure not related to tests.

File: pkg/component/delete/delete_test.go
Comment (feloy) [inline pkg/component/delete/delete_test.go:150]: @kadel do we want this behaviour, or do we want to delete the component regardless of the reult of the preStopEvents commands?

File: pkg/component/delete/delete_test.go
Comment (feloy) [inline pkg/component/delete/delete_test.go:1]: Great coverage, thanks

File: pkg/odo/cli/delete/component.go
Comment (feloy) [inline pkg/odo/cli/delete/component.go:51]: Could you create the context only when name is ""?
AS a metter of fact, it is not possible to use `genericclioptions.New` outside an odo directory, as it will fail due to the missing env file, and the context is not necessary anymore in this case ([see https://github.com/redhat-developer/odo/pull/5511/files](https://github.com/redhat-developer/odo/pull/5511/files#diff-1a997d7c41a88e0f0001418a841d488fe47beffdbfc7e2616dc52925568e5913R42).

File: tests/integration/devfile/cmd_delete_test.go
Comment (feloy) [inline tests/integration/devfile/cmd_delete_test.go:42]: could you indicate that the deployment (or component?) will be deleted?

File: tests/integration/devfile/cmd_delete_test.go
Comment (feloy) [inline tests/integration/devfile/cmd_delete_test.go:56]: could you indicate the component is deleted?

**Review by feloy — CHANGES_REQUESTED** [review]

File: pkg/odo/cli/delete/component.go
Comment (feloy) [inline pkg/odo/cli/delete/component.go:50]: Even if the language supports this, it is a better practice to explicitely return the value, to make it easier to read:
```
return err
```

File: pkg/odo/cli/delete/component.go
Comment (feloy) [inline pkg/odo/cli/delete/component.go:126]: Having this function `printDevfileComponents` to also compute `undeploy` is confusing, as these 2 things are non related from the caller point of view.

I would instead have a separate function to check if there are Kubernetes components in a devfile, that would be more eloquent

**Review by feloy — COMMENTED** [review]
Just a little remark, and a latest test is missing because of the capitalization of an error message:
```
✗  The current directory does not represent an odo component.
          To start editing your component, use "odo dev" and open this folder in your favorite IDE. Changes will be directly reflected on the cluster.
          To deploy your component to a cluster use "odo deploy".
          Or switch to directory with a component.
          
      to contain substring
          <string>: the current directory does not represent an odo component
```

File: pkg/odo/cli/delete/component.go
Comment (valaparthvi) [reply pkg/odo/cli/delete/component.go:126]: Yeah, that makes more sense. On it.

File: pkg/component/delete/delete_test.go
Comment (kadel) [reply pkg/component/delete/delete_test.go:367]: Delete should happen even if the event fails. Otherwise, it would be easy to get into a situation where it is impossible to delete the component due to a bug/type/mistake in PreStop event. 

If this turns out to be a problem, we can consider changing this in the future, to not delete if it fails, and add flag `--ignore-even-failures`.
To keep things simple for now, lets just ignore prestop events failures by default .

File: pkg/component/delete/delete_test.go
Comment (kadel) [reply pkg/component/delete/delete_test.go:386]: we definitely want to delete the component even if pre stop even fails. more info in https://github.com/redhat-developer/odo/pull/5500#discussion_r824572155

**Comment by kadel** [comment]
I would update output a little bit, in its current form
```
The following resources will be deleted: 
 •  Devfile Component: nodejs-prj1-api-abhz
 •  Kubernetes Resources:
         Deployment : my-component
         Deployment : my-component-2
? Are you sure you want to delete these resources? (y/N)
```
it looks like The devfile component and Kubernetes resources are two separate things. But in fact, the Devfile component is represented by those resources.


Something like this might be better:
```
This will delete nodejs-prj1-api-abhz from "namespace".
The component contains the following resources that will get deleted:
 - Deployment : my-component
 - Deployment : my-component-2

? Do you want to delete "component-from-devfile" and all its resources? (y/N):
```

**Comment by valaparthvi** [comment]
The output for delete with devfile is a little different than without devfile. Without devfile displays deployments matching the name and its associated endpoints that will be deleted, while this one doesn't show endpoints, is that alright @kadel ?

**Comment by kadel** [comment]
> The output for delete with devfile is a little different than without devfile. Without devfile displays deployments matching the name and its associated endpoints that will be deleted, while this one doesn't show endpoints, is that alright @kadel ?

I haven't checked that to be honest. We need to make sure that the outputs are the same.

**Comment by kadel** [comment]
> > The output for delete with devfile is a little different than without devfile. Without devfile displays deployments matching the name and its associated endpoints that will be deleted, while this one doesn't show endpoints, is that alright @kadel ?
> 
> I haven't checked that to be honest. We need to make sure that the outputs are the same.

please try to make the output look like what I suggested. We will have to update the output for deletion scenario.

**Comment by valaparthvi** [comment]
>please try to make the output look like what I suggested. We will have to update the output for deletion scenario.

Yup, I'm already on it.

File: pkg/odo/cli/delete/component.go
Comment (kadel) [inline pkg/odo/cli/delete/component.go:105]: Why are you not resuing `ListResourcesToDelete` (https://github.com/redhat-developer/odo/blob/daa9c61a4108dbef439842a17f993ebda74a8fcd/pkg/component/delete/delete.go#L26)?

**Review by kadel — CHANGES_REQUESTED** [review]
I think that this needs to be refactored to fit it with https://github.com/redhat-developer/odo/pull/5511 There is a lot of overlapping logic, that is duplicity implemented. For example all output and user confirmation should be shared between those two scenarios.

File: pkg/odo/cli/delete/component.go
Comment (feloy) [reply pkg/odo/cli/delete/component.go:105]: I'm not sure this is what we want to do here.

- With `ListResourcesToDelete`, we will find in the cluster all the resources belonging to a component, based on their labels.
- When deleting using a devfile, I understood that we want to delete the resources found in the devfile.

Comment (kadel) [reply pkg/odo/cli/delete/component.go:105]: @feloy that is a good point. I forgot about that. It makes sense to have two separate logics to handle this

**Comment by valaparthvi** [comment]
@kadel I've tried to ensure similar outputs for both delete operations, but there is some UI improvements that can be done, e.g. adding spinners here and there. I intend to leave that to @cdrage, because he's good with that. For now, everything seems to be functional.

I haven't reused the functions from #5511 because I would like to treat both the delete operations as separate, and also as Philippe mentioned in his https://github.com/redhat-developer/odo/pull/5500#discussion_r824778966.

**Comment by kadel** [comment]
I've a simple devfile with outer loop and inner loop config (https://github.com/kadel/devfile-nodejs-deploy)

I do `odo dev` than `odo deploy` and then `odo delete component`
The output shows only one Deployment and one Service.
```
$ odo delete component
This will delete "devfile-nodejs-deploy" from the namespace "myproj".
 •  The component contains the following resources that will get deleted:
	- Deployment : devfile-nodejs-deploy
	- Service : devfile-nodejs-deploy
? Are you sure you want to delete "devfile-nodejs-deploy" and all its resources? Yes
```

But there are two of each.
```
$ k get deployment,svc
NAME                                        READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/devfile-nodejs-deploy       1/1     1            1           42s
deployment.apps/devfile-nodejs-deploy-app   1/1     1            1           70s

NAME                                TYPE           CLUSTER-IP     EXTERNAL-IP   PORT(S)          AGE
service/devfile-nodejs-deploy       LoadBalancer   10.217.5.93    <pending>     3000:32466/TCP   41s
service/devfile-nodejs-deploy-app   ClusterIP      10.217.5.180   <none>        3000/TCP         70s
```

I would expect all 4 resources listed in the output of what will be deleted.

**Comment by kadel** [comment]
I also just noticed that the name of the Deployment that is being shown is not correct. I haven't checked the code yet, but it looks like it is showing the name of the component instead of the name of the Deployment.

**Comment by valaparthvi** [comment]
> I've a simple devfile with outer loop and inner loop config (https://github.com/kadel/devfile-nodejs-deploy)
> 
> I do `odo dev` than `odo deploy` and then `odo delete component` The output shows only one Deployment and one Service.
> 
> ```
> $ odo delete component
> This will delete "devfile-nodejs-deploy" from the namespace "myproj".
>  •  The component contains the following resources that will get deleted:
> 	- Deployment : devfile-nodejs-deploy
> 	- Service : devfile-nodejs-deploy
> ? Are you sure you want to delete "devfile-nodejs-deploy" and all its resources? Yes
> ```
> 
> But there are two of each.
> 
> ```
> $ k get deployment,svc
> NAME                                        READY   UP-TO-DATE   AVAILABLE   AGE
> deployment.apps/devfile-nodejs-deploy       1/1     1            1           42s
> deployment.apps/devfile-nodejs-deploy-app   1/1     1            1           70s
> 
> NAME                                TYPE           CLUSTER-IP     EXTERNAL-IP   PORT(S)          AGE
> service/devfile-nodejs-deploy       LoadBalancer   10.217.5.93    <pending>     3000:32466/TCP   41s
> service/devfile-nodejs-deploy-app   ClusterIP      10.217.5.180   <none>        3000/TCP         70s
> ```
> 
> I would expect all 4 resources listed in the output of what will be deleted. Or at least both Deployments to show. `devfile-nodejs-deploy-app` might be tricker to show.

The resources `deployment.apps/devfile-nodejs-deploy-app` and `service/devfile-nodejs-deploy-app` belong to the innerloop I assume, that were created when you ran `odo dev` for "devfile-nodejs-deploy` component.

The above output is expected, since for `odo delete` with access to devfile.yaml, we only examine the devfile, we do not touch the cluster until the actual deletion operation.

>I also just noticed that the name of the Deployment that is being shown is not correct. I haven't checked the code yet, but it looks like it is showing the name of the component instead of the name of the Deployment.

Correct, it shows the devfile component name.

**Comment by kadel** [comment]
I've discussed https://github.com/redhat-developer/odo/pull/5500#issuecomment-1066626285 with @valaparthvi The agreement is that it will be updated so it shows the list of the resources that odo is going to delete (what Kubernetes resources odo is going to api `Delete` action)

For https://github.com/kadel/devfile-nodejs-deploy it should end up looking like this

```
$ odo delete component
This will delete "devfile-nodejs-deploy" from the namespace "myproj".
 •  The component contains the following resources that will get deleted:
	- Deployment : devfile-nodejs-deploy
	- Deployment : devfile-nodejs-deploy-app
	- Service : devfile-nodejs-deploy
? Are you sure you want to delete "devfile-nodejs-deploy" and all its resources? Yes
```

File: tests/integration/devfile/cmd_delete_test.go
Comment (anandrkskd) [inline tests/integration/devfile/cmd_delete_test.go:13]: Shouldn't this be `Describe` instead?

Comment (valaparthvi) [reply tests/integration/devfile/cmd_delete_test.go]: Yup, that was left by mistake.

**Comment by valaparthvi** [comment]
@feloy @kadel 
This is up for review again. I haven't fixed unit tests yet because I feel as if I've over worked them with this PR, I'll fix them once you both approve the changes.

File: pkg/component/delete/delete.go
Comment (feloy) [inline pkg/component/delete/delete.go:100]: It would be important to use a centralized function wrapping the `fmt.Sprintf("%s-%s", componentName, appName)`, that is also used during deployment creation, so we can change the name of the deployment easily.

As far as I can see, it is `pkg.util.NamespaceKubernetesObject` that is used during deployment creation. Please use this function, and feel free to change its name to make it more expressive. You could also place it in the `pkg/component` package

File: pkg/component/delete/delete.go
Comment (feloy) [inline pkg/component/delete/delete.go:102]: Instead of returning labels to be used to delete the deployment by labels, you should return the deployment as part of the resources list, so it is identified by its kind/name, and not by its labels anymore.

I think it is dangerous to identify the deployment to delete by its labels, as you are not 100% sure there won't have other deployment with the same labels.

Going this way, you would have to change `DeleteDevfileComponent` to be some kind of `PreDeleteDevfileComponent` which will run the hooks before to delete the component, but won't delete the deployment itself (as it will be deleted as part of the call to `DeleteResources`)

**Review by feloy — CHANGES_REQUESTED** [review]

File: pkg/component/delete/delete.go
Comment (valaparthvi) [reply pkg/component/delete/delete.go:102]: >Instead of returning labels to be used to delete the deployment by labels, you should return the deployment as part of the resources list, so it is identified by its kind/name, and not by its labels anymore.

I initially had that thought, but I thought we didn't want to return a low layer k8s object to the CLI layer. Also, I didn't see a way to delete the deployment by using the deployment object itself.

>I think it is dangerous to identify the deployment to delete by its labels, as you are not 100% sure there won't have other deployment with the same labels.

I don't disagree to this, but I thought you and Tomas were specific about deleting the Deployment, which I assume to mean using the `kubeclient.Delete` function that I am currently using. So, I went ahead with fetching resource from the cluster and using it's label.

>Going this way, you would have to change DeleteDevfileComponent to be some kind of PreDeleteDevfileComponent which will run the hooks before to delete the component, but won't delete the deployment itself (as it will be deleted as part of the call to DeleteResources)

To use Delete Resources, I'd have to 1) fetch the deployment by name - this returns an `appsv1.Deployment` object,2)Convert the Deployment object into unstructured.Unstructured object with marshaling and un-marshaling - this seemed like an unnecessary task which I can't remember why.
But, if you are okay with using `DeleteResources` instead of `kubeclient.Delete(labels)`, I can make that work.

File: pkg/component/delete/delete.go
Comment (valaparthvi) [reply pkg/component/delete/delete.go:100]: I saw `pkg.util.NamespaceKubernetesObject` yday, but it seemed to be doing more than just appending the two names, so I decided to not use it, but sure I can use it.

File: pkg/component/delete/delete.go
Comment (feloy) [inline pkg/component/delete/delete.go:156]: To avoid nested blocks, you can return at this point and remove the following `else`

File: pkg/odo/cli/delete/component.go
Comment (feloy) [inline pkg/odo/cli/delete/component.go:108]: should be `isInnerLoopDeployed`

**Review by feloy — COMMENTED** [review]
I find the flow much clearer with these changes. Thanks
Just some more nit-picking ;)

**Comment by kadel** [comment]
/approve
the code needs some cleanup and tests. But functionality and output looks good to me

File: pkg/component/delete/delete.go
Comment (feloy) [reply pkg/component/delete/delete.go:116]: You could write a sub-function for this block, to be able to return ASAP in particular cases, for example the case where `IsNotFound(err) is true`.

The `err == nil` is difficult to interpret here, it should be better to have `deployment != nil` instead.

Comment (valaparthvi) [reply pkg/component/delete/delete.go]: I cannot check `deployment != nil` because it's never the case, I have settled for `deployment.Name != ""` instead.
I tried moving the block to a sub-function, but that made things worse with `isInnerLoopDeployed` involved, so I've kept things as they are.

**Pull request merged**
